### Day 5：Pandas 進階（Data Scientist 核心）⭐⭐⭐⭐⭐
這一天非常重要。

學會後，你就可以開始做真正的資料分析（EDA）。

In [2]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

## Part 1：groupby() ⭐⭐⭐⭐⭐
groupby 是 Data Scientist 最常用的功能之一。

In [6]:
#依照性別分組在計算年齡平均
df.groupby("sex")["age"].mean()

sex
female    27.915709
male      30.726645
Name: age, dtype: float64

In [ ]:
#依照艙等分組計算平均票價
df.groupby("class")["fare"].mean()

class
First     84.154687
Second    20.662183
Third     13.675550
Name: fare, dtype: float64

In [8]:
#依照性別計算生還率
df.groupby("sex")["survived"].mean()

sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

In [9]:
#依照艙等計算平均年齡
df.groupby("class")["age"].mean()

class
First     38.233441
Second    29.877630
Third     25.140620
Name: age, dtype: float64

In [10]:
#依照embark_town計算平均價格
df.groupby("embark_town")["fare"].mean()

embark_town
Cherbourg      59.954144
Queenstown     13.276030
Southampton    27.079812
Name: fare, dtype: float64

### Part 2：多個 groupby ⭐⭐⭐⭐⭐
可以同時依照兩個欄位分組。

In [13]:
#不同性別 × 艙等的平均年齡。
df.groupby(["sex","class"])["age"].mean()

sex     class 
female  First     34.611765
        Second    28.722973
        Third     21.750000
male    First     41.281386
        Second    30.740707
        Third     26.507589
Name: age, dtype: float64

In [12]:
#不同性別 × 艙等的生還率。
df.groupby(["sex", "class"])["survived"].mean()

sex     class 
female  First     0.968085
        Second    0.921053
        Third     0.500000
male    First     0.368852
        Second    0.157407
        Third     0.135447
Name: survived, dtype: float64

### Part 3：pivot_table() ⭐⭐⭐⭐⭐
pivot table 可以把 groupby 的結果變成表格。

In [14]:
pd.pivot_table(
    df,
    values = "survived",
    index = "sex",
    columns = "class"
)

class,First,Second,Third
sex,,,
female,0.968085,0.921053,0.500000
male,0.368852,0.157407,0.135447


In [27]:
#建立不同性別 × 艙等的平均年齡表。
pd.pivot_table(
    df,
    values = "age",
    index = "sex",
    columns = "class"
)

class,First,Second,Third
sex,,,
female,34.611765,28.722973,21.750000
male,41.281386,30.740707,26.507589


In [29]:
#建立同性別 × 艙等的平均票價表。
pd.pivot_table(
    df,
    values = "fare",
    index = "sex",
    columns = "class"
)

class,First,Second,Third
sex,,,
female,106.125798,21.970121,16.118810
male,67.226127,19.741782,12.661633


In [30]:
#建立不同 embark_town × class 的平均生還率。
pd.pivot_table(
    df,
    values = "survived",
    index = "embark_town",
    columns = "class"
)

class,First,Second,Third
embark_town,,,
Cherbourg,0.694118,0.529412,0.378788
Queenstown,0.500000,0.666667,0.375000
Southampton,0.582677,0.463415,0.189802


### Part 4：缺失值（Missing Value）⭐⭐⭐⭐⭐

In [ ]:
#先查看缺失值
df.isnull()
#統計每隔欄位缺失值數量
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

# 哪個欄位缺失值最多？

deck 有688個缺失值

### Part 5：dropna() ⭐⭐⭐⭐⭐

In [33]:
#刪除缺失值
df.dropna()


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True
10,1,3,female,4.0,1,1,16.7000,S,Third,child,False,G,Southampton,yes,False
11,1,1,female,58.0,0,0,26.5500,S,First,woman,False,C,Southampton,yes,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
871,1,1,female,47.0,1,1,52.5542,S,First,woman,False,D,Southampton,yes,False
872,0,1,male,33.0,0,0,5.0000,S,First,man,True,B,Southampton,no,True
879,1,1,female,56.0,0,1,83.1583,C,First,woman,False,C,Cherbourg,yes,False
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True


In [36]:
#刪除age缺失值的新DataFrame
df2 = df.dropna(subset = ["age"])

In [ ]:
#比較刪除前與刪除後的資料量
print(len(df))
print(len(df2))

891
714


### Part 6：fillna() ⭐⭐⭐⭐⭐

In [39]:
#使用平均值填補並建立新欄位
df["age_fill"] = df["age"].fillna(
    df["age"].mean()
)

In [40]:
#利用平均票價填補缺失值。
df["fare_fill"] = df["fare"].fillna(
    df["fare"].mean()
)

In [41]:
#確認新欄位缺失值數量
df["age_fill"].isnull().sum()


np.int64(0)

### Part 7：merge() ⭐⭐⭐⭐⭐
Data Scientist 幾乎每天都在合併資料。

In [43]:
student = pd.DataFrame({
    "ID" : [1,2,3],
    "Name" : ["Amy", "Bob", "Cindy"]
})

score = pd.DataFrame({
    "ID" : [1,2,3],
    "Score" : [90,80,95]
})
#合併兩個dictionary
pd.merge(student, score)

,ID,Name,Score
0,1,Amy,90
1,2,Bob,80
2,3,Cindy,95


In [44]:
fruit = pd.DataFrame({
    "ID" : [1,2,3],
    "Product" : ["Apple", "Banana", "Orange" ]
})

price = pd.DataFrame({
    "ID" : [1,2,3],
    "Price" : [30, 20, 25]
})
pd.merge(fruit, price)

,ID,Product,Price
0,1,Apple,30
1,2,Banana,20
2,3,Orange,25


In [4]:
name = pd.DataFrame({
    "ID" : [101,102,103],
    "Name" : ["Tom", "Mary", "John"]
})

depart = pd.DataFrame({
    "ID" : [101,102,103],
    "Department" : ["HR", "Finance", "IT"]
})

pd.merge(name, depart)

,ID,Name,Department
0,101,Tom,HR
1,102,Mary,Finance
2,103,John,IT


### Mini Project 1 ⭐⭐⭐⭐

分析：

不同性別的平均年齡與生還率。

In [5]:
df.groupby("sex")[["age", "survived"]].mean()

,age,survived
sex,,
female,27.915709,0.742038
male,30.726645,0.188908


### Mini Project 2 ⭐⭐⭐⭐⭐

分析：

不同艙等的：

平均年齡
平均票價
生還率

In [7]:
df.groupby("class")[["age", "fare", "survived"]].mean()

,age,fare,survived
class,,,
First,38.233441,84.154687,0.629630
Second,29.877630,20.662183,0.472826
Third,25.140620,13.675550,0.242363


### Mini Project 3 ⭐⭐⭐⭐⭐

建立 Pivot Table：

列：sex

欄：class

數值：survived

觀察：「哪一群人最容易生還？」

In [9]:
pd.pivot_table(
    df,
    values = "survived",
    index = "sex",
    columns = "class"
)

class,First,Second,Third
sex,,,
female,0.968085,0.921053,0.500000
male,0.368852,0.157407,0.135447


### Bonus Project ⭐⭐⭐⭐⭐

「女性是否比男性更容易生還？」

「First Class 是否比 Third Class 更容易生還？」

In [20]:
sex_survival = df.groupby("sex")["survived"].mean()
sex_survival.sort_values(ascending= False) #在新的表格中排序


sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

In [24]:
class_survival = df.groupby("class")["survived"].mean()
class_survival.sort_values(ascending= False)


class
First     0.629630
Second    0.472826
Third     0.242363
Name: survived, dtype: float64

## 女性是否比男性更容易生還？ 

女生的平均生還率有74%，遠高於男生的19%，女生比男生更容易生還，因此可推測女性乘客在鐵達尼號事故中明顯較容易獲得救援機會。

## First Class 是否比 Third Class 更容易生還？

Fitrst Class的平均生還率有63%，Third Class的平均生還率僅24%，因此可推知在艙等越高的旅客，生還率也越高。

## Final Findings

本次分析發現：

1.女性的生還率顯著高於男性。

2.艙等越高，生還率越高。

3.First Class 為生還率最高的艙等。

4.Third Class 為生還率最低的艙等。

## 綜合結果顯示：

性別與艙等都與 Titanic 乘客的生還機率具有明顯關聯性。